rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,516 @@
+
+Service levels are specific to the user's targets.  While it's impossible to
+properly measure whether or not an end-user's app is meeting service level
+targets, we can establish a predefined histogram that would allow Couchbase to
+simply and inexpensively understand if the service level has likely changed.
+
+It's proposed that we establish a set of standardized histograms, specific to
+services and to the service level expected with the service using those
+parameters.  Counters for these histograms can be sent to Couchbase and both
+stored individually and aggregated. See histogram buckets in "Histograms" table
+below.
+
+As an aggregated set of metrics, it would give Couchbase a rough view over the
+service level the user is experiencing.  Deviations in service level can be dug
+into and, in particular, we can check for correlation to Couchbase Capella
+management activities like upgrades.
+
+The Indexing Service is not included as it is not used by SDK, even for
+management operations. All future services must be reported if the service
+exposes separate port and the SDK is using it.
+
+As with the Service Experience Counters, it may be incorrect to try to
+compare between deployments.  For example, the query complexity may differ
+significantly or the data size and use of features like compression may vary
+significantly between two applications.
+
+The SDK must record latency for histogram on the lower layer for each
+transmission attempt instead of top-level operation. For example, if KV GET
+operation results in three network requests, there should be three data points
+for the metrics counter, and if the operation has not been canceled, it should
+record three data points for histogram, even if the operation has been timed
+out, because on the lower-level API, the SDK still sees orphan responses and is
+able to figure out true latency of the request.
+
+# API and Implementation Details
+
+The decision to reuse the existing Meter implementation is considered an SDK
+implementation detail, and this RFC does not specify the API for collecting
+metrics from the library.
+
+One possible approach is to introduce a "compound meter" that aggregates an
+existing meter (which could be overridden by the user) and the application
+service-level telemetry meter described in this RFC. This compound meter would
+collect counters and fixed-bucket latency histograms. The application
+service-level meter would continue to function even if the user provides a
+custom meter implementation.
+
+In addition to latencies, which would be used for reporting histograms, the
+default Meter should also maintain the following counters:
+
+| Metric | Description |
+| :---- | :---- |
+| sdk\_*{service}*\_r\_timedout | Timeout |
+| sdk\_*{service}*\_r\_canceled | Canceled |
+| sdk\_*{service}*\_r\_total | Total number of operations |
+
+The SDK must send zero metric, because it still carries information that at
+given timestamp, no events have been registered.
```

Comment:
Should we be sending zero metrics for _all_ services, with _all_ possible `node_uuid`/`node`/`bucket` combinations? So if, for example, so far we've only seen KV operations should we also send zero metrics for the 5 other services? That would end up being a large number of zero metrics.

Reply:
no, it does not make sense to send zero metrics (although I think we still want to send histogram buckets with zero hits)